text/0000-generalize-over-mut-and-non-mut-items.md

```diff
@@ -0,0 +1,175 @@
+    fn opt-baz(&mut self) -> &opt i32;
+    fn qux(&self) -> i32;
+    fn mut-tux(&self) -> i32 { 42 }
+}
+```
+
+Code snippet #2:
+```
+struct Thing<'a> {
+    x:     &'a i32,
+    y: &'a mut i32,
+    z:         i32,
+}
+
+struct mut-Thing<'a> {
+    x: &'a mut i32,
+    y: &'a mut i32,
+    z:         i32,
+}
+
+fn foo(thing: &Thing) -> &i32 {
+    &thing.x
+}
+
+fn mut-foo(thing: &mut Thing) -> &mut i32 {
+    &mut thing.x
+}
+
+// Notice that the qux method, which was not affected by the opt keyword
+// and not prefixed by `mut-`, is placed under this non-mut trait, and that
+// the `mut-tux` method is not placed under here.
+trait Trait {
+    fn bar(&self) -> &Self;
+    fn baz(&mut self) -> &i32;
+    fn qux(&self) -> i32;
+}
+
+// Notice that the qux method, which was not affected by the opt keyword
+// and not prefixed by `mut-`, is not placed under this mut qualified trait,
+// and that the `mut-tux` method is placed under here.
+trait mut-Trait {
+    fn mut-bar(&mut self) -> &mut Self;
+    fn mut-baz(&mut self) -> &mut i32;
+    fn mut-tux(&self) -> i32 { 42 }
+}
+```
+
+Given either code snippet #1 or #2 (both are equivalent), you could then implement separately either `Trait` or `mut-Trait` or both for a type just like you normally would (just remember that in the `mut-` version of the trait, the methods are named `mut-whatever` instead of `whatever`). But you could also implement the `Trait` and `mut-Trait` traits for a certain type both at once by using the syntax introduced in code snippet #3.
+
+Code snippet #3:
+```
+struct S { x: i32 }
+
+impl opt-Trait for S {
+    fn opt-bar(&opt self) -> &opt S {
+        self
+    }
+
+    fn opt-baz(&mut self) -> &opt i32 {
+        &opt self.x
+    }
+
+    fn qux(&self) -> i32 {
+        self.x
+    }
+}
+```
+
+What's further, not only could you implement two traits for a single type at once (as demonstrated in code snippet #3), but you could also implement two traits for two types at once. That is, you could implement for example `Trait` for `Thing` and `mut-Trait` for `mut-Thing` at once (so, it doesn't implement one trait per each type but rather does a 1-to-1 mapping). This is demonstrated in the following code snippet #4.
+
+Code snippet #4:
+```
+// Just to recap, this is what the original opt-Thing looked like:
+struct opt-Thing<'a> {
+    x: &'a opt i32,
+    y: &'a mut i32,
+    z:         i32,
+}
+
+impl opt-Trait for opt-Thing {
+    fn opt-bar(&opt self) -> &opt opt-Thing {
+        self
+    }
+
+    fn opt-baz(&mut self) -> &opt i32 {
+        self.x
+    }
+
+    fn qux(&self) -> i32 {
+        self.z
+    }
+}
+```
+
+You can call those `opt-` and `mut-` prefixed methods just like any other functions. Just think of the prefix as part of the function name:
+
+Code snippet #5:
+```
+trait opt-OtherTrait {
+    fn opt-other_foo(&mut self);
+}
+
+impl<T> opt-OtherTrait for T
+    where T: opt-Trait
+{
+    fn opt-other_foo(&mut self) {
+        let t = self.opt-bar();
```

Comment:
@reem Who do you mean it's ambiguous to?
* The parser
* The compiler
* Programmer who is reading quickly and doesn't have syntax highlighting

It shouldn't be ambiguous to the first two, because the `opt` keyword followed by one or more white space followed by `-` would always be a compile-time error.